### PR TITLE
feat(toggle): on/off icons for toggle

### PR DESCRIPTION
**Pull request by sean-perkins**
## Pull request checklist

Please check if your PR fulfills the following requirements:
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been reviewed and added / updated if needed (for bug fixes / features)
  - Some docs updates need to be made in the `ionic-docs` repo, in a separate PR. See the [contributing guide](https://github.com/ionic-team/ionic-framework/blob/main/.github/CONTRIBUTING.md#modifying-documentation) for details.
- [x] Build (`npm run build`) was run locally and any changes were pushed
- [x] Lint (`npm run lint`) has passed locally and any fixes were made for failures

## Pull request type

 

 

Please check the type of change your PR introduces:
- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe): 

## What is the current behavior?

The Toggle component does not support the accessibility indications for the toggle checked state.

Issue URL: #20524

## What is the new behavior?

Introduces the `switchLabelsEnabled` property to `ion-toggle`, that developers can use to enable the icon indicators for the toggle's checked state.

For Material Design, the checkbox icon will display when the toggle is checked. When the toggle is unchecked, the minus icon will display.

For iOS, the "I" icon will display when the toggle is checked. When the toggle is unchecked, the "O" icon will display. iOS renders both icons in the DOM, in order to accomplish the pulling effect when the toggle changes state. 

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

## Other information

Dev-build: `6.1.10-dev.11655149274.1a5ec2d4`

### CODE REVIEW

File: core/src/components/toggle/toggle.tsx [hunk 1]
```diff
@@ -204,7 +224,15 @@ export class Toggle implements ComponentInterface {
       >
         <div class="toggle-icon" part="track">
           <div class="toggle-icon-wrapper">
-            <div class="toggle-inner" part="handle" />
+            {switchLabelsEnabled && mode === 'ios' && (
+              <ion-icon class="toggle-switch-icon checked" icon={this.getSwitchLabelIcon(true)}></ion-icon>
+            )}
+            <div class="toggle-inner" part="handle">
+              {switchLabelsEnabled && mode === 'md' && this.renderOnOffSwitchLabels()}
```
Comment (liamdebeasi) [inline core/src/components/toggle/toggle.tsx:231]: Optional: We can pass in the mode to both `getSwitchLabelIcon` and `renderOnOffSwitchLabels` so we do not need to re-call it again each time the function is called.

Comment (liamdebeasi) [inline core/src/components/toggle/toggle.tsx:231]: Also any reason why this is extracted to a function but the iOS icons are written inline?

File: core/src/components/toggle/toggle.tsx [hunk 2]
```diff
@@ -63,6 +64,11 @@ export class Toggle implements ComponentInterface {
    */
   @Prop() value?: string | null = 'on';
 
+  /**
+   * Enables the on/off accessibility switch labels within the toggle.
+   */
+  @Prop() switchLabelsEnabled = false;
```
Comment (liamdebeasi) [inline core/src/components/toggle/toggle.tsx:70]: APIs should be present tense according to our style guide.

Maybe something like `enableLabels`, `enableOnOffLabels`, or `enableSwitchLabels`? I am partial to "enableOnOffLabels" as there is no confusion as to whether or not this prop impacts `ion-label`.

Comment (liamdebeasi) [reply core/src/components/toggle/toggle.tsx:70]: What do you think about defaulting this to `undefined`? If Capacitor adds the ability to automatically detect the I/O preference in the device settings, we could support that functionality without breaking changes as developers would be able to opt-out of this functionality by explicitly setting `false`. If we default to `false`, then having the feature be enabled by default is harder to do without breaking changes.

I know the design doc states it should default to `false`, so feel free to disregard. Just might make life a bit easier down the road if we support that new functionality.

**Review by liamdebeasi — CHANGES_REQUESTED** [review]
The "-" does not show up in dark mode: 
image

Also the checkmark text appears black. We should probably add some screenshot tests for dark mode.

**Review by liamdebeasi — COMMENTED** [review]
Also the I/O label supposed to appear, not slide in on iOS

| native | ionic |
| ------ | ----- |
| <video src="https://user-images.githubusercontent.com/2721089/173444653-83815949-7e9a-477b-844e-821d31dea3cf.MP4"></video> | <video src="https://user-images.githubusercontent.com/2721089/173444693-e6c6cf1d-843a-4f8c-a269-086b0cf57269.MP4"></video> |

----

Somewhat related, but when dragging the toggle the "O" label overlaps the handle

| native | ionic |
| ------ | ----- |
| <video src="https://user-images.githubusercontent.com/2721089/173444959-664ed21f-0cf5-4d5d-b9c8-c00599e02111.MP4"></video> | <video src="https://user-images.githubusercontent.com/2721089/173444991-cbafd9ca-6db8-4b80-bf76-cf723f3aad9f.MP4"></video> |

File: core/src/components/toggle/toggle.tsx (hunk 2 above)
Comment (sean-perkins) [reply core/src/components/toggle/toggle.tsx:70]: I think defaulting to undefined makes sense here, I'll update the design doc.

**Comment by sean-perkins** [comment]
Pushed an update to fix the colors for iOS to align with the spec. This does cause an issue with `ion-toggle` when `color="dark"` and the toggle is checked:

Screen Shot 2022-06-13 at 5 26 13 PM

Unsure if we should concern ourself with overriding the text color for this, since it isn't within the iOS spec.

I'll need to do some more discovery into rendering of the icons. Originally when just having the icon flip on iOS (like it does for MD), it made the animation of the knob look choppy. It looks like when the knob is active, I need to set the opacity of the icons to zero (and make sure the icons have a transition for opacity).

**Comment by liamdebeasi** [comment]
When rendering the on icon over an element with an Ionic color, it might be easiest to use the contrast (ex: `var(--ion-color-primary-contrast)` color. That should work for both light and dark mode.

For the off text, you might be able to use a step color? The color would change slightly across light vs. dark mode, but I don't know if it really matters (as long as the text is accessible). A benefit of using a stepped color is it is easy for developers to customize on a global level. Full customization is out of scope for this task, but using a stepped color would at least let developers customize the text color without much hassle. (For example, if a developers customized the track color, our default color may not be accessible when used on the new track color)

**Review by amandaejohnston — CHANGES_REQUESTED** [review]

**Review by amandaejohnston — APPROVED** [review]
Nice work!

File: core/src/components/toggle/toggle.ios.vars.scss [hunk 3]
```diff
@@ -75,3 +75,9 @@ $toggle-ios-item-end-padding-bottom:   5px !default;
 
 /// @prop - Padding start of the toggle positioned on the end in an item
 $toggle-ios-item-end-padding-start:    $item-ios-padding-start !default;
+
+/// @prop - The text color of the on/off labels
+$toggle-ios-on-off-label-color: #b0b0b0 !default;
```
Comment (liamdebeasi) [inline core/src/components/toggle/toggle.ios.vars.scss:80]: The color on light mode is not accessible:

Screen Shot 2022-06-15 at 2 09 01 PM

Dark mode seems fine though 

Screen Shot 2022-06-15 at 2 10 09 PM

File: core/src/components/toggle/toggle.md.vars.scss [hunk 4]
```diff
@@ -99,3 +99,9 @@ $toggle-md-item-end-padding-bottom:               12px !default;
 
 /// @prop - Padding start of the toggle positioned on the end in an item
 $toggle-md-item-end-padding-start:                $item-md-padding-start !default;
+
+/// @prop - The text color of the on/off labels
+$toggle-md-on-off-label-color: #000 !default;
```
Comment (liamdebeasi) [inline core/src/components/toggle/toggle.md.vars.scss:104]: These colors seem good -- we will likely need to expose a way to customize these colors in the future though

**Comment by sean-perkins** [comment]
The change from the `sizes` test is from this change:
```css
.toggle-icon-wrapper {
  align-items: center;
}
```

|Before|After|
|---|---|
|Screen Shot 2022-06-15 at 4 08 25 PM| Screen Shot 2022-06-15 at 4 08 06 PM|

I can see the argument for the before, since the shadow is contained within the toggle. I'll explore alternative patterns to vertically align the on/off label icons without using flex box.

**Comment by sean-perkins** [comment]
Ok, I think this is finally ready for another review. Was getting burned by trying to nest our SASS mixins:
```scss
@include ltr() {
  @include position(x, x, x, x);
}
```
Instead of just using the positioning values directly and disabling stylelint.

File: core/src/components/toggle/toggle.md.vars.scss (hunk 4 above)
Comment (liamdebeasi) [reply core/src/components/toggle/toggle.md.vars.scss:104]: How does the color get set for the "O" on iOS mode?

This definitely some scope creep, so feel free to ignore this for now... But what if we had a global css var for this?

For example, tab bar has a default background that can be customized with `--ion-tab-bar-background`: https://github.com/ionic-team/ionic-framework/blob/1b1b1a3800c4d044b4a3e7418f534e9271770ec6/core/src/themes/ionic.theme.default.md.scss#L17

Reason why I bring it up is iOS does change the on/off text color when adding high contrast mode.

File: core/src/components/toggle/toggle.tsx [hunk 5]
```diff
@@ -178,8 +184,29 @@ export class Toggle implements ComponentInterface {
...
+  };
+
+  private renderOnOffSwitchLabels(mode: Mode, checked: boolean) {
+    const icon = this.getSwitchLabelIcon(mode, checked);
+
+    return (
+      <ion-icon
+        class={{
+          'toggle-switch-icon': true,
+          'toggle-switch-icon--checked': checked,
```
Comment (liamdebeasi) [inline core/src/components/toggle/toggle.tsx:201]: Minor thing, but this convention does not appear to be used anywhere else in the framework. Can we make this decision as a team before adopting this syntax?

**Review by liamdebeasi — COMMENTED** [review]
The "O" doesn't seem to be centered

image

**Comment by sean-perkins** [comment]
Screen Shot 2022-07-05 at 3 30 48 PM

The "O" label appears to be vertically centered. On iOS it is not horizontal aligned (it appears to be offset further to the left than the right). The screenshot is at 150% zoom.

Your screenshot appears to draw the SVG asset thicker 🤔

**Review by liamdebeasi — COMMENTED** [review]
Another thought: It might be better to do some more design work around this CSS Variable idea we've been playing with. I'd be fine merging this feature in for now, and then we can figure out the CSS Variable stuff next sprint. Does that work for you? I think this feature is great and pretty much done, so I don't think it makes sense to hold this up on the CSS Variable detail.

**Comment by sean-perkins** [comment]
Yeah I can revert the CSS variable changes and leave them static for now.

**Review by liamdebeasi — APPROVED** [review]
Much better. Great job!

**Pull request merged**